test/heartbeat_test.cpp

```diff
@@ -0,0 +1,160 @@
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <cinttypes>
+
+#include "rclcpp/logging.hpp"
+#include "rclcpp/node.hpp"
+#include "rclcpp/executors.hpp"
+
+#include "interfaces/msg/heartbeat.hpp"
+
+#include "common_package/common_node.hpp"
+
+TEST(common_package, heartbeat_rate)
+{
+    constexpr uint16_t heartbeat_period = 500;
+    rclcpp::executors::SingleThreadedExecutor executor;
+
+    std::shared_ptr<CommonNode> heartbeat_node = std::make_shared<CommonNode>("heartbeat");
+    rclcpp::Node::SharedPtr test_node = std::make_shared<rclcpp::Node>("test");
+
+    interfaces::msg::Heartbeat last_msg;
+    last_msg.tick = 0;
+    last_msg.time_stamp = test_node->now();
+
+    rclcpp::Subscription<interfaces::msg::Heartbeat>::SharedPtr heartbeat_sub = test_node->create_subscription<interfaces::msg::Heartbeat>(
+        "heartbeat",
+        1,
+        [&last_msg, test_node](interfaces::msg::Heartbeat::ConstSharedPtr msg)
+        {
+            RCLCPP_DEBUG(test_node->get_logger(), "Got message with tick: %" PRIu32, msg->tick);
+            ASSERT_TRUE(rclcpp::Time(msg->time_stamp) - last_msg.time_stamp <= rclcpp::Duration(std::chrono::duration<int64_t, std::milli>(heartbeat_period + 10)));
+            ASSERT_EQ(last_msg.tick + 1, msg->tick);
+            ASSERT_FALSE(msg->active);
+            ASSERT_EQ(msg->sender_id, "/heartbeat");
+            last_msg = *msg; });
+
+    rclcpp::TimerBase::SharedPtr end_timer = test_node->create_wall_timer(
+        std::chrono::milliseconds(heartbeat_period * 10), [&executor, test_node]()
+        {
+            RCLCPP_DEBUG(test_node->get_logger(), "Stopping executor");
+            executor.cancel(); });
+
+    executor.add_node(heartbeat_node);
+    executor.add_node(test_node);
+
+    executor.spin();
+    ASSERT_EQ(last_msg.tick, 10);
+}
+
+TEST(common_package, heartbeat_blocked)
```

Comment:
Der Test ist so leider nicht aussagekräftig.
Nach 500ms wird der erste Heartbeat gesendet und von dem Callback empfangen.
Nach 750ms triggered der block timer und blockt für 2500ms.
Danach ruft er cancel auf. Damit werden keine weiteren Callbacks getriggered.
Das bedeutet auch, das eventuell anliegende Heartbeats in der Queue vom Subscriber nicht mehr bearbeitet werden.
Streng genommen wird nach außen hin also nur gezeigt, das der Subscriber geblockt wird.
Um eine Aussage treffen zu können, dass wirklich der Publisher geblockt wurde, darf der Subscriber nicht mitgeblockt werden.

Reply:
Gut, dann hab ich da deinen ursprünglichen Test nicht verstanden. Muss ich mir nochmal in Ruhe anschauen.

Reply:
Kann ihn dir sonst auch später mal erklären.